Refactor Laser Profiles to Functors

ax3l opened the pull request:
Refactor laser profile creation:

- cleanup confusing includes of LaserPhysics.hpp
- remove `laserManipulation` from `FieldE` and `FieldB`
- move laser Kernel to `LaserPhysics`

Builds the basis for future features such as

- multiple lasers
- lasers from various planes / various directions
- pitched laser to target normal

## To Do

- [x] rebase on top of #2609

### Refactor Laser Profiles

- [x] None (*untested*)
- [x] GaussianBeam (*RT test with LWFA*)
- [x] PulseFrontTilt (*untested*)
- [x] PlaneWave (*RT test with LWFA*)
- [x] Wavepacket (*untested*)
- [x] Polynom (*untested*)
- [x] ExpRampWithPrepulse (*untested*)

### Refactor Laser Examples

- [x] Bremsstrahlung (*untested*)
- [x] Bunch (*untested*)
- [x] FoilLCT (*untested*)
- [x] LaserWakefield

And remove `components.param` in examples that use the default options.

### Link Lasers in Docs as Section

- [x] link each laser's default params
  - [x] default params documented in default `laser.param` or in `.def` file?
  - [x] move/add docstrings to `.def` file of each profile or to default param in `laser.param`?

ax3l commented:
@PrometheusPi @steindev @wmles @BeyondEspresso I updated the laser profiles from its old legacy code base to a new, clean implementation.

Can you assist me reviewing and testing our laser profiles, please?

PrometheusPi commented:
@ax3l of course - I will have a look at it

ax3l commented on include/picongpu/fields/laserProfiles/Polynom.hpp:
```diff
@@ -0,0 +1,266 @@
+
+            float_X const exponent =
+                ( pos.x() / Unitless::W0x ) * ( pos.x() / Unitless::W0x ) +
+                ( pos.z() / Unitless::W0z ) * ( pos.z() / Unitless::W0z );
```
needs 2D generalization

psychocoderHPC replied:
opened #2602 to support the generalization

ax3l commented on include/picongpu/fields/laserProfiles/PulseFrontTilt.hpp:
```diff
@@ -0,0 +1,305 @@
+            // planeNoNormal[ planeNormalDir ] = 0.0;
+            // Gaussian Beam with zero tilt:
+            //            r2 = math::abs2( pos * planeNoNormal );
+            float_X const r2 = ( pos.x() + spaceShift ) * ( pos.x() + spaceShift ) + pos.z() * pos.z();
```
needs 2D generalization

ax3l commented on include/picongpu/fields/laserProfiles/Wavepacket.hpp:
```diff
@@ -0,0 +1,267 @@
+
+            // transversal position only
+            float_X const exp_x = pos.x() * pos.x() / ( Unitless::W0_X * Unitless::W0_X );
+            float_X const exp_z = pos.z() * pos.z() / ( Unitless::W0_Z * Unitless::W0_Z );
```
needs 2D generalization

psychocoderHPC commented on include/picongpu/fields/laserProfiles/Wavepacket.hpp:
```diff
@@ -0,0 +1,267 @@
+            typename T_Acc
+        >
+        HDINLINE acc::Wavepacket< Unitless >
+        operator()(
```
Please rebase this PR: Operator must be const since the interface change in #2600

ax3l commented on include/picongpu/param/laser.param:
```diff
@@ -473,22 +578,12 @@ namespace picongpu
-        }
-    }
+    //! currently selected laser profile
+    using Profile = None<>;
```
Possibly rename `Profile` to `Selected`

steindev commented on include/picongpu/param/laser.param:
```diff
@@ -21,15 +21,95 @@
+ *  - None                : no laser init
+ *  - GaussianBeam        : Gaussian beam (focusing)
+ *  - PulseFrontTilt      : Gaussian beam with a tilted pulse envelope
+ *                          in 'x' direction
```
Better describe which axes enclose the pulse-front tilt angle. At the moment the description is ambiguous.

ax3l commented on include/picongpu/fields/LaserPhysics.hpp:
```diff
@@ -17,83 +17,196 @@
+                 *          A second order solver could need 2 neighbors left:
+                 *            -> initialize in y=0 and y=1 plane
+                 *
+                 * Question: Why do other codes initialize the B-Field instead?
```
That's the same as is written there.

ax3l replied:
> always first state what you do

I agree. The title of the comment block (L154) is:

> Calculate how many neighbors to the left we have to initialize the laser in the E-Field

sbastrakov replied:
I also find this long comment rather confusing, but for a different reason.

To me the issue is with `Because our fields are defined on the lower cell side (C-Style ftw)`. Each Yee cell by being a cell and not a node has two values for each field component. So it's not that the fields are defined on the lower side, but the indexing of nodes and cells in picongpu is organized in such a way. Also both ways of organizing Yee cells (E on edges, B on faces or vice versa) are popular, so that might potentially add to reader's confusion.

Maybe it's useful (or it even exists already, but I am unaware) to have a picture in documentation which explains how components are organized in a Yee cell and also what indexing is used.

ax3l replied:
Good point! We can add some in a follow up to the manual, I have them in my diploma thesis: https://zenodo.org/record/15924

ax3l commented on include/picongpu/fields/laserProfiles/ExpRampWithPrepulse.def:
```diff
@@ -0,0 +1,178 @@
+{
+namespace defaults
+{
+    struct ExpRampWithPrepulseParam
```
This file will probably never be seen by a frontend user.
It's just about adding proper docs in Doxygen so we can use the doc string in the manual.

steindev replied:
Agreed. I was already wondering why the information seems to be repeated in `laser.param`

ax3l commented on include/picongpu/fields/laserProfiles/Polynom.def:
```diff
@@ -0,0 +1,129 @@
+        /** unit: Volt / meter */
+        static constexpr float_64 AMPLITUDE_SI = 1.738e13;
+
+        /** The profile of the test Lasers 0 and 2 can be stretched by a
```
This is some weird, arcane old comment indeed. will use the same doc string as in other laser profiles.

PrometheusPi commented:
@ax3l Could you please fix the parenthesis error in `picongpu/fields/laserProfiles/Polynom.hpp(167)` - so I can test the polynomial laser profile.

ax3l commented:
Must have destroyed it yesterday before going to bed :)

ax3l commented:
@PrometheusPi @BeyondEspresso @wmles any test results for your refactored lasers? :)

psychocoderHPC commented:
@ax3l can I merge this PR, I would like to avoid that this is very long open and must be  rebased to often.

@PrometheusPi  @BeyondEspresso  @wmles  Do you started the testing and if not what is the expected time scale where you can test it.

PrometheusPi commented:
Gauss Laser looks Okay
![image](https://user-images.githubusercontent.com/5121158/40723436-36bffb28-641f-11e8-8991-b5d24a22becb.png)

PrometheusPi commented on include/picongpu/fields/laserProfiles/PulseFrontTilt.def:
```diff
@@ -0,0 +1,135 @@
+
+        /** Polarization selection
+         */
+        static constexpr PolarisationType Polarisation = CIRCULAR;
```
Just a remark: the previous default was `LINEAR_X`. If you want to keep the previous behavior, please keep it linear polarized.

ax3l replied:
@PrometheusPi saw your table below, will keep it `LINEAR_X` then!

PrometheusPi commented:
The tilted laser looks Okay:

![image](https://user-images.githubusercontent.com/5121158/40725659-4a9a9162-6424-11e8-8bf0-74ccfad6d45f.png)

PrometheusPi commented:
The plane wave laser agrees perfectly:

![image](https://user-images.githubusercontent.com/5121158/40770428-e6c41392-64ba-11e8-8fe3-6291c6540f6e.png)

PrometheusPi commented:
I see a major difference in the wave packet laser

![image](https://user-images.githubusercontent.com/5121158/40784002-8a5fc950-64e4-11e8-9ed8-58167ae43ca8.png)

I will check further. A first recheck of the parameters revealed no difference.

PrometheusPi commented:
There seems to be an issue with the initialization of the wave packet laser.

This is the on-axis field after 1000 iterations.

![image](https://user-images.githubusercontent.com/5121158/40784400-dcd48cce-64e5-11e8-96ce-3f369842d880.png)

PrometheusPi commented on include/picongpu/fields/laserProfiles/Wavepacket.hpp:
```diff
@@ -0,0 +1,270 @@
+        static constexpr float_X AMPLITUDE = float_X( Params::AMPLITUDE_SI / UNIT_EFIELD ); // unit: Volt /meter
+        static constexpr float_X W0_X = float_X( Params::W0_X_SI / UNIT_LENGTH ); // unit: meter
+        static constexpr float_X W0_Z = float_X( Params::W0_Z_SI / UNIT_LENGTH ); // unit: meter
+        static constexpr float_X INIT_TIME = float_X( ( Params::PULSE_INIT * Params::PULSE_LENGTH_SI ) / UNIT_TIME ); // unit: seconds (full initialization length)
```
This definition ignores the constant region. Therefore, the laser initialization starts in the center of the pulse. 
Old code:
```
constexpr float_X INIT_TIME = float_X ((RAMP_INIT*SI::PULSE_LENGTH_SI + SI::LASER_NOFOCUS_CONST\
ANT_SI) / UNIT_TIME); // unit: seconds (full inizialisation length)  
```
as with the plane wave laser.

PrometheusPi commented on include/picongpu/fields/laserProfiles/Polynom.hpp:
```diff
@@ -0,0 +1,270 @@
+            // @todo add half-cells via traits::FieldPosition< Solver::NumicalCellType, FieldE >()
+
+            // transversal position only
+            float3_X const w0_3D( Unitless::W0_X, 0., Unitless::W0_Z );
```
The naming of these variables differ a bit from the other lasers. In `Unitless::` they are named `W0x` and `W0z`.

PrometheusPi commented:
The exp laser looks fine 

![image](https://user-images.githubusercontent.com/5121158/40839266-0d068764-65a2-11e8-8fd8-7fb790e6a5d9.png)

ax3l commented:
@PrometheusPi thank you very much for the review! I updated the missing constant are in the init of the wavepacket laser and fixed the inconsistent variable names in the polynomial laser profile.

PrometheusPi commented:
@ax3l Great! Do you want to keep the new polarization of the pulse front tilt laser?

ax3l commented:
@PrometheusPi yes, let's keep this in the default so it's consistent with the other lasers' polarization.

PrometheusPi commented:
The update did not seem to fix the issue with the wave packet laser 
![image](https://user-images.githubusercontent.com/5121158/40968425-cb8f7ea0-68b5-11e8-9495-6126da883fc7.png)

ax3l commented:
@PrometheusPi ok, I will then keep all the lasers' polarization as is.

ax3l commented:
@PrometheusPi fixed the last commit (amended): forgot a 0.5 init_time shift in the wavepacket.

PrometheusPi commented:
I still see a slight difference in the polynom laser

![image](https://user-images.githubusercontent.com/5121158/40971453-f903142a-68bd-11e8-8186-49cca96ce85b.png)


and that is not coming from the different parameter mentioned below.

PrometheusPi commented on include/picongpu/fields/laserProfiles/Polynom.def:
```diff
@@ -0,0 +1,128 @@
+         *  Info:             FWHM_of_Intensity = FWHM_Illumination
+         *                      = what a experimentalist calls "pulse duration"
+         *  unit: seconds (1 sigma) */
+        static constexpr float_64 PULSE_LENGTH_SI = 10.615e-15 / 4.0;
```
In the previous version, this has just been `4.0e-15`.

PrometheusPi commented on include/picongpu/fields/laserProfiles/Polynom.def:
```diff
@@ -0,0 +1,128 @@
+
+        /** Stretch temporal profile by a constant plateau between the up and downramp
+         *  unit: seconds */
+        static constexpr float_64 LASER_NOFOCUS_CONSTANT_SI = 13.34e-15;
```
This variable is not used in the code of the polynomial laser.

PrometheusPi commented on include/picongpu/fields/laserProfiles/Polynom.def:
```diff
@@ -0,0 +1,128 @@
+        /** The laser pulse will be initialized PULSE_INIT times of the PULSE_LENGTH
+         *
+         *  unit: none */
+        static constexpr float_64 PULSE_INIT = 20.0;
```
This is not used.

PrometheusPi commented on include/picongpu/fields/laserProfiles/Polynom.hpp:
```diff
@@ -0,0 +1,270 @@
+        static constexpr float_X AMPLITUDE = float_X( Params::AMPLITUDE_SI / UNIT_EFIELD ); // unit: Volt /meter
+        static constexpr float_X W0_X = float_X( Params::W0_X_SI / UNIT_LENGTH ); // unit: meter
+        static constexpr float_X W0_Z = float_X( Params::W0_Z_SI / UNIT_LENGTH ); // unit: meter
+        static constexpr float_X INIT_TIME = float_X( ( Params::PULSE_INIT * Params::PULSE_LENGTH_SI ) / UNIT_TIME ); // unit: seconds (full initialization length)
```
An init parameter is not needed for the polynomial laser.
The old code was:
```c++
constexpr float_X INIT_TIME = float_X (SI::PULSE_LENGTH_SI / UNIT_TIME);
```
Since the laser has **no** exponential reduction in amplitude but is zero outside its pulse length.

psychocoderHPC commented on include/picongpu/fields/laserProfiles/Polynom.hpp:
```diff
@@ -0,0 +1,270 @@
+            const float_X t_rise = 0.5 * Unitless::PULSE_LENGTH;
+            const float_X tau = runTime / t_rise;
+
+            const float_X omegaLaser = 2.0 * PI * Unitless::f;
```
must be `float_x( 2.0 ) * ...`

psychocoderHPC commented on include/picongpu/fields/laserProfiles/Polynom.hpp:
```diff
@@ -0,0 +1,270 @@
+             * making the laser pulse 2*t_rise long
+             */
+
+            const float_X t_rise = 0.5 * Unitless::PULSE_LENGTH;
```
must be `float_X( 0.5 ) * ...`

psychocoderHPC commented on include/picongpu/fields/laserProfiles/Polynom.hpp:
```diff
@@ -0,0 +1,270 @@
+        Tpolynomial( float_X const tau )
+        {
+            float_X result( 0.0 );
+            if( tau >= 0.0 && tau <= 1.0 )
```
please cast all scalars to `float_X`

ax3l replied:
Omg, this block is madness xD

PrometheusPi commented:
The wave packet laser looks (after the last amended update) fine. 

![image](https://user-images.githubusercontent.com/5121158/40975183-6509b2ee-68ca-11e8-8fc3-5a448e2fb527.png)

The difference is the most severe one but it is orders of magnitude below the amplitudes. 

We might need to check the wave packet laser again, because the plateau does not look like a plateau but more like a ramp:
![image](https://user-images.githubusercontent.com/5121158/40975266-a7e2365e-68ca-11e8-885d-6baf714d9cf1.png)
(both for the `dev` and the current pull request)

PrometheusPi commented on include/picongpu/fields/laserProfiles/ExpRampWithPrepulse.def:
```diff
@@ -0,0 +1,177 @@
+        static constexpr float_64 AMPLITUDE_SI = _A0 * UNITCONV_A0_to_Amplitude_SI;
+
+        /** unit: Volt /meter */
+        //constexpr float_64 AMPLITUDE_SI = 1.738e13;
```
@ax3l This is equal to an a_0 of 4.33 but above a a_0 of 20 is used. Should we make this consistent with this pull request or do we leave it as is/as it was?

ax3l replied:
If you like we can make them consistent in a follow-up.

PrometheusPi commented:
Now, the polynomial laser looks fine. :+1:

![image](https://user-images.githubusercontent.com/5121158/40981502-51cc08c4-68db-11e8-8f89-3896f2de728e.png)

PrometheusPi commented on include/picongpu/fields/LaserPhysics.hpp:
```diff
@@ -17,83 +17,198 @@
+     *
+     * @tparam T_numWorkers number of workers
+     * @tparam T_LaserPlaneSizeInSuperCell number of cells per dimension which
+     *  initialize the laser (size must be less or equal than the supercell size)
```
Optional:
less than or equal to

PrometheusPi commented on include/picongpu/fields/LaserPhysics.hpp:
```diff
@@ -17,83 +17,198 @@
+        void operator()(uint32_t currentStep) const
+        {
+            /* The laser can be initialized in the plane of the first cell or
+             * any later x-y plane inside the simulation. Initializing the
```
Question:
Is it not a x-z plane with a specific y coordinate/cell-value?

psychocoderHPC replied:
yes must be `x-z`

PrometheusPi commented on include/picongpu/fields/LaserPhysics.hpp:
```diff
@@ -17,83 +17,198 @@
+            /* The laser can be initialized in the plane of the first cell or
+             * any later x-y plane inside the simulation. Initializing the
+             * laser in planes inside the simulation corresponds to an
+             * evaluation of the field at negatively shifted time.
```
This is not clear

PrometheusPi commented on include/picongpu/fields/LaserPhysics.hpp:
```diff
@@ -17,83 +17,198 @@
+                 *
+                 * Example: Yee needs one neighbor to perform dB = curlE
+                 *            -> initialize in y=0 plane
+                 *          A second order solver could need 2 neighbors left:
```
The word `left` seems misplaced.

PrometheusPi replied:
Do you mean on the left side?

PrometheusPi commented on include/picongpu/fields/laserProfiles/ExpRampWithPrepulse.def:
```diff
@@ -0,0 +1,177 @@
+        static constexpr float_X INT_RATIO_POINT_2 = 1.e-4;
+        static constexpr float_X INT_RATIO_POINT_3 = 1.e-4;
+
+        // time-positions of prepulse and preramps points
```
adding a 
```c++
/** unit: seconds */
```
would be helpful and consistent with the rest of the file

PrometheusPi commented on include/picongpu/fields/laserProfiles/ExpRampWithPrepulse.def:
```diff
@@ -0,0 +1,177 @@
+         *  Info:             FWHM_of_Intensity = FWHM_Illumination
+         *                      = what a experimentalist calls "pulse duration"
+         *  unit: seconds (1 sigma) */
+        static constexpr float_64 PULSE_LENGTH_SI = 3.0e-14 / 2.35482; // half of the time in which E falls to half its initial value (then I falls to half its value in 15fs, approx 6 wavelengths). Those are 4.8 wavelenghts.
```
This is much longer than 80 characters

PrometheusPi commented on include/picongpu/fields/laserProfiles/ExpRampWithPrepulse.def:
```diff
@@ -0,0 +1,177 @@
+     *
+     * Allows defining a prepulse and two regions of exponential preramp with
+     * independent slopes. The definition works by specifying three (t, intensity)-
+     * points, where time is counted from the very beginning in SI and the
```
The addition `in SI` is over specific

PrometheusPi commented on include/picongpu/fields/laserProfiles/ExpRampWithPrepulse.hpp:
```diff
@@ -0,0 +1,396 @@
+
+        static constexpr float_X WAVE_LENGTH = float_X( Params::WAVE_LENGTH_SI / UNIT_LENGTH ); // unit: meter
+        static constexpr float_X PULSE_LENGTH = float_X( Params::PULSE_LENGTH_SI / UNIT_TIME ); // unit: seconds (1 sigma)
+        static constexpr float_X LASER_NOFOCUS_CONSTANT = float_X( Params::LASER_NOFOCUS_CONSTANT_SI / UNIT_TIME ); // unit: seconds
```
longer than 80 characters

PrometheusPi replied:
several times in this file

ax3l commented:
@kossag14 @codingS3b careful: on the next update you pull in from `dev`, the `laser.param` syntax has changed.
